pkg/ipcache/metadata.go

```diff
@@ -100,11 +100,11 @@ func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTr
 	)
 
 	idMDMU.Lock()
-	defer idMDMU.Unlock()
```

Comment:
Looking at `InjectLabels()` in isolation, I think this would be OK because `InjectLabels()` is not intended to be reentrant, and there's only one caller anyway so the code inside doesn't need to protect against multiple callers calling in concurrently.

However, `InjectLabels()` has a corresponding opposite operation, `RemoveLabelsExcluded()` which has similar logic and these two will currently mutually exclude each other from concurrently manipulating the IdentityMetadata map + apply the changes resulting from those changes. This exclusion occurs using `idMDMU`, which would change with the current proposal in this PR. This adds additional complications to resolving the overall issue for two reasons:
1. Presumably if this issue is caused by policy calculation (potentially for any FQDN policy) happening at the same time as `InjectLabels()` (from node / k8s apiserver updates), similarly an equivalent issue could also occur if policy calculation happens at the same time that `RemoveLabelsExcluded()` occurs (this time, only from k8s apiserver updates). So while this solves one of the problems, there is still another similar issue in a related part of the code.
2. Regardless of the problem that is highlighted in the PR description,  it will now also make it possible for one goroutine to be handling e.g. a node update inside `InjectLabels()` while another goroutine handles a kube-apiserver endpoint update via `RemoveLabelsExcluded()` -> `removeLabelsFromIPs()`. This is not currently possible because the `idMDMU` mutex puts a critical section around the core logic for both of these functions. With the current proposal, if `InjectLabels()` runs slightly earlier, then it will grab the `idMDMU` lock and manipulate the IdentityMetadata map and calculate the changes that must subsequently occur, then release the lock, and then depending on the scheduling of goroutines on the system, could plausibly be delayed arbitrarily while the `RemoveLabelsExcluded()` call grabs the idMDMU, manipulates & calculates new state, and then issues the updates into the SelectorCache & IPCache. Depending on the scheduling, any of the possible state transitions could occur:
   a. All of `InjectLabels()` occurs, then all of `RemoveLabelsExcluded()`
   b. All of `RemoveLabelsExcluded()` occurs, then all of `InjectLabels()`
   c. The first part of `InjectLabels()` occurs here, then due to rescheduling of goroutines, all of `RemoveLabelsExcluded()` occurs, then the remainder of `InjectLabels()` occurs.

(2) starts to get a bit abstract, probably if the IPs are entirely different then there's no problem. But if the IPs collide between the two implementations, then rather than consistently first applying the changes to the `IdentityMetadata` map and then deriving the resulting state to apply to the `SelectorCache` / `IPCache` in order, it could end up mixing the order of operations and I'm not entirely convinced whether the code would end up consistently getting the SelectorCache+IPcache into the same state regardless of which conditions (a), (b), or (c) occur.

Broadly I see a couple of options that could resolve both of these issues: One is to ensure that both `InjectLabels()` and `RemoveLabelsExcluded()` occur inside the same goroutine, or the second would be to extend _some_ lock to cover the critical sections of both of these functions. My initial thought is "Well what about the `IPIdentityCache.mutex`?". Unfortunately, looking over the callstacks you provided, this will only defer the deadlock until slightly later since the other callstack will also attempt to grab the `IPIdentityCache.mutex` and push those changes down to the datapath while also holding the SelectorCache mutex.

Reply:
For the immediate term, I wonder whether we can simply add a brand new mutex specifically for coordinating just `InjectLabels()` vs. `RemoveLabelsExcluded()`. Either function would first grab this `ApplyIDMDChanges` mutex before the `idMDMU`, and release this mutex only once all of the IdentityMetadata, SelectorCache and IPCache operations are completed. Then it should be safe to shorten the scope of the current IDMDU lock exactly the way that this patch is already proposing, and additionally we could apply a similar fix for the equivalent `RemoveLabelsExcluded()` code.

Longer term, we probably need to restructure the handling of events in and around the IPcache subsystem, and the way it interacts with the policy / `SelectorCache` layers to formalize the ordering of operations and provide better guarantees. But I don't have a fully formed idea around what that would look like just yet, and it's probably a bit unreasonable to immediately push for such a solution from you given that you've only recently started digging around in the code :-)

Reply:
Good point. Thinking about this again, I didn't realize the `RemoveLabelsExcluded()` angle. 

From point (2) above, anything other than option (a) is undesirable (or vice versa where `RemoveLabelsExcluded()` completes first, then `InjectLabels()`; of course assuming that the K8s events are in order, but I think that's a safe assumption).

I agree in the immediate term a separate mutex makes since which stitches together the two other mutexes, eliminating the gap in critical section, so-to-speak.